File: src/game/TLeft.ts
import { TSide } from './TSide';
import { TBoard } from './TBoard';
import { GAME_CONSTANTS } from '../constants/GameConstants';

export class TLeft extends TSide {
    constructor(hDC: CanvasRenderingContext2D, game: TBoard) {
        super(game, GAME_CONSTANTS.DL);
        
        // Initialize position matching the original
        this.sColor = this.hGame.getCell(this.hGame.getH() - 1, 0) - 1;
        this.hGame.setCell(this.hGame.getH() - 1, 0, this.playerCode);
        this.score = 1;
        this.check(hDC, 0);
        this.sColor = 0;
        this.change(hDC, 0);
    }

    check(hDC: CanvasRenderingContext2D, forbid: number): number {
        let i: number, j: number, ii: number;
        let backi: boolean, backj: boolean;

        if (this.sColor !== this.fieldColor - 1 && this.sColor !== forbid - 1) {
            this.fieldColor = this.sColor + 1;
        } else {
            return 0;
        }

        i = this.hGame.getH() - 1;
        while (i >= 0) {
            backi = false;
            j = 0;
            while (j < this.hGame.getW()) {
                backj = false;
                ii = i - 1 + (2 * (j % 2));
                
                if (this.hGame.getCell(i, j) & this.playerCode) {
                    this.drawRomb(hDC, i, j, this.fieldColor - 1);
                    
                    if (ii >= 0 && ii < this.hGame.getH()) {
                        if (j + 1 < this.hGame.getW()) {
                            if (this.hGame.getCell(ii, j + 1) === this.fieldColor) {
                                this.hGame.setCell(ii, j + 1, this.playerCode);
                                this.score++;
                                if (ii > i) backi = true;
                            }
                        }
                        if (j - 1 >= 0) {
                            if (this.hGame.getCell(ii, j - 1) === this.fieldColor) {
                                this.hGame.setCell(ii, j - 1, this.playerCode);
                                this.score++;
                                backj = true;
                                if (ii > i) backi = true;
                            }
                        }
                    }
                    
                    if (j + 1 < this.hGame.getW()) {
                        if (this.hGame.getCell(i, j + 1) === this.fieldColor) {
                            this.hGame.setCell(i, j + 1, this.playerCode);
                            this.score++;
                        }
                    }
                    if (j - 1 >= 0) {
                        if (this.hGame.getCell(i, j - 1) === this.fieldColor) {
                            this.hGame.setCell(i, j - 1, this.playerCode);
                            this.score++;
                            backj = true;
                        }
                    }
                    
                    this.hGame.resetCell(i, j, this.playerCode + this.fieldColor);
                }
                
                if (!backj) j++;
                else if (j) j--;
            }
            if (!backi) i--;
            else if (i < this.hGame.getH() - 1) i++;
        }

        this.printScore(hDC);
        
        if (this.score > (this.hGame.getW() * this.hGame.getH() - this.hGame.getW() / 2) / 2) {
            this.animation(hDC);
            return -1;
        } else {
            return 1;
        }
    }

    printScore(hDC: CanvasRenderingContext2D): void {
        hDC.fillStyle = 'rgb(255,255,255)';
        hDC.font = '16px Arial';
        const scoreText = `${this.score}    `;
        hDC.fillText(scoreText, GAME_CONSTANTS.WIDTH / 4 - GAME_CONSTANTS.RECT_S, GAME_CONSTANTS.HEIGHT + 50);
    }

    animation(hDC: CanvasRenderingContext2D): void {
        // Implementation for win animation
        // This would match the original animation logic
    }

    // Override positioning methods for left side
    protected getRectColLeft(): number {
        return GAME_CONSTANTS.WIDTH / 4 - GAME_CONSTANTS.RECT_S * 2;
    }

    protected getRectColTop(): number {
        return GAME_CONSTANTS.HEIGHT + 20; // Position in the extra space we added
    }

    protected getRectColRight(): number {
        return this.getRectColLeft() + GAME_CONSTANTS.RECT_S / 2;
    }

    protected getRectColBottom(): number {
        return this.getRectColTop() + GAME_CONSTANTS.RECT_S / 2;
    }
} 

File: CPP/TLeft.cpp
// TLeft.cpp: implementation of the TLeft class.
//
//////////////////////////////////////////////////////////////////////

#include "stdafx.h"
#include "TLeft.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

TLeft::TLeft(HDC *hDC,TBoard *Game)
{
	hGame = Game;
	playerCode = 64;

	rectCol.left= WIDTH/4 - RECT_S*2;
	rectCol.top = HEIGHT-93;
	rectCol.right = rectCol.left + RECT_S/2;
	rectCol.bottom = rectCol.top + RECT_S/2 ;

	rect.left = rectCol.left - RECT_S/4;
	rect.top = rectCol.top - RECT_S/4;
	rect.right = rectCol.right + RECT_S/4;
	rect.bottom = rectCol.bottom + RECT_S/4 ;

	SColor = hGame->GetCell(hGame->GetH()-1,0) - 1;
	hGame->SetCell(hGame->GetH()-1,0,playerCode);
	Score = 1;
	Check(hDC, 0);
	SColor = 0;
	Change(hDC,0);
}

TLeft::~TLeft()
{

}

int TLeft::Check(HDC *hDC,int forbid)
{
	int i, j, ii;
	bool backi, backj;
	if( SColor != FieldColor-1 && SColor != forbid-1)
		FieldColor = SColor + 1;
	else return 0;
	i = hGame->GetH() - 1;
	while( i >= 0)
	{  backi = false;
		j = 0;
		while( j < hGame->GetW() )
		{  backj = false;
			ii = i-1+(2*(j%2));
			if( hGame->GetCell(i,j) & playerCode )
			{  DrawRomb(hDC,i,j, FieldColor - 1);
				if( ii >= 0 &&  ii < hGame->GetH())
				{	if( j+1 < hGame->GetW() )
						if( hGame->GetCell(ii,j+1) == FieldColor )
						{	hGame->SetCell(ii, j+1, playerCode);
							Score++;
							if( ii > i ) backi = true;}
					if( j-1 >=0 )
						if( hGame->GetCell(ii,j-1) == FieldColor )
						{	hGame->SetCell(ii,j-1, playerCode);
							Score++;
							backj = true;
							if( ii > i ) backi = true; }
				}
				if( j + 1 < hGame->GetW() )
					if( hGame->GetCell(i,j+1) == FieldColor )
					{	hGame->SetCell(i,j+1, playerCode);
						Score++; }
						if( j - 1 >=0 )
				if( hGame->GetCell(i, j-1) == FieldColor )
					{	hGame->SetCell(i, j-1, playerCode);
						Score++;
						backj = true; }
				hGame->ResetCell(i,j, playerCode + FieldColor);
			}
			if( !backj ) j++;
			else if( j ) j--;
		}
		if( !backi ) i--;
		else if ( i < hGame->GetH()-1 ) i++;
	}
	PrintScore(hDC);
	if (Score > (hGame->GetW() * hGame->GetH() - hGame->GetW()/2)/2 )
	{ 	Animation(hDC);
		return  -1;	}
	else return  1;
}

void TLeft::PrintScore(HDC *hDC)
{
	char temp[10];
	SetTextColor( *hDC, RGB (255,255,255) );
	SetBkColor( *hDC, RGB (0,0,0) );
	wsprintf(temp, "%d%s", Score, "    ");
	TextOut( *hDC, WIDTH/4 - RECT_S, HEIGHT-70, temp, strlen(temp));

}

void  TLeft::Animation(HDC *hDC)
{
	int i,j;
	for(j = 0; j < hGame->GetW(); j++)
		for(i = hGame->GetH()-1; i >=0; i--)
			if( hGame->GetCell(i, j) )
			{	DrawRomb(hDC,i,j, FieldColor-1);
				hGame->ResetCell(i, j, FieldColor);
			}
}

